thunx-encoding-json/src/main/java/com/contentgrid/thunx/encoding/json/ExpressionJsonConverter.java

```diff
@@ -75,8 +76,66 @@ public JsonExpressionDto visit(Variable variable) {
             return new JsonVariableDto(variable.getName());
         }
 
+        @Override
+        protected JsonExpressionDto visit(CollectionValue collectionValue) {
+            return processCollectionDto(collectionValue);
+        }
+    }
+
+    private static class JsonScalarEncoderVisitor extends ContextFreeThunkExpressionVisitor<JsonScalarDto<?>> {
+
+        @Override
+        protected JsonScalarDto<?> visit(Scalar<?> scalar) {
+            return processScalar(scalar);
+        }
+
+        @Override
+        protected JsonScalarDto<?> visit(FunctionExpression<?> functionExpression) {
+            throw new UnsupportedOperationException("Only Scalars are supported in Collection");
+        }
+
+        @Override
+        protected JsonScalarDto<?> visit(SymbolicReference symbolicReference) {
+            throw new UnsupportedOperationException("Only Scalars are supported in Collection");
+        }
+
+        @Override
+        protected JsonScalarDto<?> visit(Variable variable) {
+            throw new UnsupportedOperationException("Only Scalars are supported in Collection");
+        }
+
+        @Override
+        protected JsonScalarDto<?> visit(CollectionValue collectionValue) {
+            return processCollectionDto(collectionValue);
+        }
+    }
 
+    private static JsonScalarDto<?> processScalar(Scalar<?> scalar) {
+        var resultType = scalar.getResultType();
+
+        var typeName = JsonScalarDto.SCALAR_TYPES.get(resultType);
+        if (typeName == null) {
+            String exMessage = String.format("Scalar of type <%s> is not supported", resultType.getName());
+            throw new IllegalArgumentException(exMessage);
+        }
+
+        return JsonScalarDto.of(typeName, scalar.getValue());
     }
 
+    private static JsonScalarDto<?> processCollectionDto(CollectionValue collectionValue) {
+        Collection<JsonExpressionDto> result;
+        if (JsonCollectionValueDto.getTypeByClass(collectionValue.getResultType()).equals("set")) {
```

Comment:
If you need to do typechecks here, this indicates that we need two separate visit methods: one for the set type and one for the array type. 

This also means that we need two separate classes: `CollectionValue` needs to be split up in `ArrayCollectionValue` and `SetCollectionValue`. You can keep `CollectionValue` as an interface so it can be used in places where an array or a set would be sufficient.

Reply:
Yeap, I agree with you, but can we merge this PR and create issue, for this? I will pick it up and fix in next PR, cause this one became to big and too time consuming. This change requires rewriting the whole PR:(